### PR TITLE
Ensure python builds uses all ressources

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

File: native_client/ctcdecode/setup.py [hunk 1]
```diff
@@ -56,12 +56,7 @@ def read(fname):
 
 decoder_module = Extension(
     name='ds_ctcdecoder._swigwrapper',
-    sources=['swigwrapper.i',
-             'ctc_beam_search_decoder.cpp',
-             'scorer.cpp',
-             'path_trie.cpp',
-             'decoder_utils.cpp',
-             'workspace_status.cc'],
```
Comment (reuben) [inline native_client/ctcdecode/setup.py]: This is done on purpose as a compromise between developer speed of iteration and faster builds. If you move everything into common.a you have to rebuild the world on every change to these files.

Comment (lissyx) [reply native_client/ctcdecode/setup.py]: > This is done on purpose as a compromise between developer speed of iteration and faster builds. If you move everything into common.a you have to rebuild the world on every change to these files.

Yes, but on ctc decoder build, I can measure ~5 min win thanks to that because we can parallelize them this way.

Maybe I can put them into a second `.a` unit ?

Comment (reuben) [reply native_client/ctcdecode/setup.py]: 5 minutes?! D:

File: native_client/ctcdecode/setup.py [hunk 2]
```diff
@@ -40,33 +40,42 @@
...
+
+        build_common(srcs=srcs,
+                     out_name=out_name,
+                     build_dir=build_dir,
+                     num_parallel=known_args.num_processes,
+                     debug=debug)
+
 project_version = read('../../VERSION').strip()
 
+CTC_DECODER_FILES = [
```
Comment (reuben) [inline native_client/ctcdecode/setup.py]: Move this to `build_common.py` along with the other constants?

File: native_client/ctcdecode/setup.py [hunk 3]
```diff
@@ -83,5 +92,5 @@ class BuildExtFirst(build):
     ext_modules=[decoder_module],
     package_dir = {'ds_ctcdecoder': '.'},
     py_modules=['ds_ctcdecoder', 'ds_ctcdecoder.swigwrapper'],
-    install_requires = ['numpy%s' % numpy_min_ver],
+    install_requires = ['numpy%s' % numpy_min_ver]
```
Comment (reuben) [inline native_client/ctcdecode/setup.py]: nit: undo this change

File: native_client/ctcdecode/build_common.py [hunk 4]
```diff
@@ -32,14 +32,14 @@
         'unittest.cc'))
 ]
 
-def build_common(out_name='common.a', build_dir='temp_build/temp_build', debug=False, num_parallel=1):
+def build_common(srcs=[], out_name='common.a', build_dir='temp_build/temp_build', debug=False, num_parallel=1):
```
Comment (reuben) [inline native_client/ctcdecode/build_common.py]: `build_common` is no longer a good name, and the default value for `out_name` also doesn't make sense.

File: native_client/ctcdecode/setup.py (hunk 2 above)
Comment (lissyx) [reply native_client/ctcdecode/setup.py]: I was unsure, so I'mfine

File: native_client/ctcdecode/build_common.py (hunk 4 above)
Comment (lissyx) [reply native_client/ctcdecode/build_common.py]: @reuben would `build_ctc.py` be fine? and `kenlm.a` for default value ?

Comment (reuben) [reply native_client/ctcdecode/build_common.py]: Just `build_archive` is fine.

Comment (reuben) [reply native_client/ctcdecode/build_common.py]: And no default value.

File: native_client/ctcdecode/setup.py (hunk 1 above)
Comment (lissyx) [reply native_client/ctcdecode/setup.py]: Yeah, full task completing in 19 min instead of 24 min.

File: native_client/ctcdecode/setup.py [hunk 5]
```diff
@@ -40,33 +40,42 @@
...
+    'decoder_utils.cpp',
+    'workspace_status.cc'
+]
 
-if not os.path.exists(common_build):
-    if not os.path.exists(build_dir):
-        os.makedirs(build_dir)
+build_dir = 'temp_build/temp_build'
+archive_build = 'archive.a'
+ctc_decoder_build = 'ctc_decoder.a'
```
Comment (reuben) [inline native_client/ctcdecode/setup.py]: These are bad names. How about `first_party.a` and `third_party.a`?

File: native_client/ctcdecode/setup.py (hunk 2 above)
Comment (reuben) [reply native_client/ctcdecode/setup.py]: Looks like you forgot to remove this.

File: native_client/ctcdecode/build_archive.py [hunk 6]
```diff
@@ -20,26 +20,34 @@
...
-def build_common(out_name='common.a', build_dir='temp_build/temp_build', debug=False, num_parallel=1):
+CTC_DECODER_FILES = [
+    'ctc_beam_search_decoder.cpp',
+    'scorer.cpp',
+    'path_trie.cpp',
+    'decoder_utils.cpp',
+    'workspace_status.cc'
+]
+
+def build_common(srcs=[], out_name='', build_dir='temp_build/temp_build', debug=False, num_parallel=1):
```
Comment (reuben) [inline native_client/ctcdecode/build_archive.py]: You renamed the file but not the function.

File: native_client/ctcdecode/Makefile [hunk 7]
```diff
@@ -17,12 +17,13 @@ endif
 
 all: bindings
 
-clean-keep-common:
+clean-keep-archive:
```
Comment (reuben) [inline native_client/ctcdecode/Makefile]: Name this `clean-keep-third-party`, and it needs to delete the first party archive.

File: native_client/ctcdecode/Makefile [hunk 8]
```diff
@@ -17,12 +17,13 @@ endif
...
 	rm -rf dist temp_build ds_ctcdecoder.egg-info
 	rm -f swigwrapper_wrap.cpp swigwrapper.py
 
-clean: clean-keep-common
-	rm -f common.a
+clean: clean-keep-archive
```
Comment (reuben) [inline native_client/ctcdecode/Makefile]: And then this one deletes the third party archive too (changes less often).

Comment (lissyx) [reply native_client/ctcdecode/Makefile]: sorry, tired and the rules are really confusing when you are constantly distracted
